scheds/rust/scx_bpfland/src/bpf/main.bpf.c

```diff
@@ -345,35 +371,59 @@ static u64 calc_avg_clamp(u64 old_val, u64 new_val, u64 low, u64 high)
 }
 
 /*
- * Compare two vruntime values, returns true if the first value is less than
- * the second one.
- *
- * Copied from scx_simple.
+ * Return a value inversely proportional to a weight.
  */
-static inline bool vtime_before(u64 a, u64 b)
+static u64 scale_inverse_fair(u64 value, u64 weight)
 {
-	return (s64)(a - b) < 0;
+	return value * 100 / weight;
 }
 
 /*
- * Return task's average amount of context switches per second.
+ * Compute the deadline component of a task (this value will be added to the
+ * task's vruntime to determine the actual deadline).
  */
-static u64 task_avg_nvcsw(struct task_struct *p)
+static s64 task_compute_dl(struct task_struct *p ,struct task_ctx *tctx)
 {
-	struct task_ctx *tctx;
+	/*
+	 * The amount of voluntary context switches contributes to determine
+	 * the task's priority.
+	 */
+	u64 task_prio = p->scx.weight + tctx->avg_nvcsw;
 
-	tctx = try_lookup_task_ctx(p);
-	if (!tctx)
+	/*
+	 * If not in "lowlatency" mode, always apply a pure vruntime based
+	 * scheduling.
+	 */
+	if (!lowlatency)
+		return 0;
+
+	/*
+	 * If the task has not ran during the previous slice_ns period, use its
+	 * vruntime as deadline to give it a priority boost. This allows to
+	 * speed up tasks that are mostly sleeping and they suddenly need to
+	 * react fast.
+	 */
+	if (vtime_before(tctx->last_running + slice_ns, bpf_ktime_get_ns()))
```

Comment:
I guess such a task is mostly synchronously wakened up. If so, does this give double prioritization in addition to SCX_WAKE_SYNC handling?

Reply:
Possibly... in this case SCX_WAKE_SYNC will promote the task to interactive, so it'll be dispatched to the priority DSQ and in addition to that the deadline will be set equal to its vruntime, so it will likely be executed before the other tasks in the prio DSQ.

This may result in excessive prioritization, but it seems to help to maintain system responsiveness when running massive hackbench / perf bench stress tests (that are also classified as interactive tasks and dispatched to the priority DSQ).